libbeat/cmd/instance/beat.go

```diff
@@ -772,6 +784,47 @@ func (b *Beat) createOutput(stats outputs.Observer, cfg common.ConfigNamespace)
 	return outputs.Load(b.index, b.Info, stats, cfg.Name(), cfg.Config())
 }
 
+func (b *Beat) registerClusterUUIDFetching() error {
+	if b.Config.Output.Name() == "elasticsearch" {
+		callback, err := b.clusterUUIDFetchingCallback()
+		if err != nil {
+			return err
+		}
+		elasticsearch.RegisterConnectCallback(callback)
+	}
+	return nil
+}
+
+// Build and return a callback to fetch the Elasticsearch cluster_uuid for monitoring
+func (b *Beat) clusterUUIDFetchingCallback() (elasticsearch.ConnectCallback, error) {
+	var response struct {
+		ClusterUUID string `json:"cluster_uuid"`
+	}
+
+	callback := func(esClient *elasticsearch.Client) error {
+		status, body, err := esClient.Request("GET", "/", "", nil, nil)
+		if err != nil {
+			return errw.Wrap(err, "error querying /")
+		}
+		if status > 299 {
+			return fmt.Errorf("Error querying /. Status: %d. Response body: %s", status, body)
+		}
+		err = json.Unmarshal(body, &response)
+		if err != nil {
+			return fmt.Errorf("Error unmarshaling json when querying /. Body: %s", body)
+		}
+
+		stateRegistry := monitoring.GetNamespace("state").GetRegistry()
+		outputsRegistry := stateRegistry.NewRegistry("outputs")
+		elasticsearchRegistry := outputsRegistry.NewRegistry("elasticsearch")
+		monitoring.NewString(elasticsearchRegistry, "cluster_uuid").Set(response.ClusterUUID)
+
+		return nil
+	}
```

Comment:
If the user configures multiple hosts, this callback will be called per host, potentially at the same time. The use of `response` is not thread-safe (even if they report the same value, race detector will complain).

Plus each call will update the registry. Not sure right now, but NewString or NewRegistry might even panic if the "key" is already known.